src/org/jitsi/impl/neomedia/rtp/TransportCCEngine.java

```diff
@@ -50,12 +65,156 @@
     private byte extensionId = -1;
 
     /**
-     * The next sequence number to use for outgoing packets.
+     * The next sequence number to use for outgoing data packets.
      */
     private AtomicInteger outgoingSeq
         = new AtomicInteger(new Random().nextInt() & 0xffff);
 
     /**
+     * The running index of sent RTCP transport-cc feedback packets.
+     */
+    private AtomicInteger outgoingFbPacketCount = new AtomicInteger();
+
+    /**
+     * The list of {@link MediaStream} that are using this
+     * {@link TransportCCEngine}.
+     */
+    private final List<MediaStream> mediaStreams = new LinkedList<>();
+
+    /**
+     * Incoming transport-wide sequence numbers mapped to the timestamp of their
+     * reception (in milliseconds since the epoch).
+     */
+    private TreeMap<Integer, Long> incomingSeqs;
+
+    /**
+     * Used to synchronize access to {@link #incomingSeqs}.
+     */
+    private final Object incomingSeqsSyncRoot = new Object();
+
+    /**
+     * The time (in milliseconds since the epoch) at which the first received
+     * packet in {@link #incomingSeqs} was received (or -1 if the map is empty).
+     * Kept here for quicker access, because the map is ordered by sequence
+     * number.
+     */
+    private long firstIncomingTs = -1;
+
+    /**
+     * Notifies this instance that a data packet with a specific transport-wide
+     * sequence number was received on this transport channel.
+     * @param seq the transport-wide sequence number of the packet.
+     * @param marked whether the RTP packet had the "marked" bit set.
+     */
+    private void packetReceived(int seq, boolean marked)
+    {
+        long now = System.currentTimeMillis();
+        synchronized (incomingSeqsSyncRoot)
```

Comment:
`ArrayBlockingQueue` has an atomic `clear`, would that work?  Seems like clearing it (rather than setting to null and re-allocating a new one) might be nice as well?